### PR TITLE
feat: 2d screenspace rendering

### LINKED ISSUE #1: Add 2D screenspace rendering
Support rendering 2D overlays: text and boxes.

This is required for the in-game console and other GUI / HUD elements.

### PR DESCRIPTION
**Pull request by michaeleggers**
closes #1 

2D font and box rendering API example:

`Begin2D()` and `End2D()` must be inside `RenderBegin()` and `RenderEnd()`

```cpp
// Enable screenspace 2D rendering. Binds the 2d offscreen framebuffer and activates the 2d shaders.
m_Renderer->Begin2D();

                                          // Font to use      // Color
m_Renderer->SetFont( m_ConsoleFont, glm::vec4(1.0f, 1.0f, 1.0f, 1.0f) );
m_Renderer->DrawText("Hello World", 0.0f, 0.0f);

// If you want to draw in absolute coordinates then you have to specify it.
// Depends on the resolution of the render window! 
m_Renderer->DrawText("Some more text in absolute coordinates :)", 0.0f, 200.0f, COORD_MODE_ABS); 

// Draw a red, opaque box that starts at the top left corner of the window and goes to the
// center of the window
m_Renderer->SetShapeColor( glm::vec4(1.0f, 0.0f, 0.0f, 1.0) );
m_Renderer->DrawBox( 0.0f, 0.0f, 0.5f, 0.5f);

// Same with shapes: If you want absolute coordinates, you have to specify it:
m_Renderer->DrawBox( 600, 600, 200, 100, COORD_MODE_ABS );

// If you want to draw text in a different size you have to use another font object.
m_Renderer->SetFont( m_ConsoleFont30, glm::vec4(0.3f, 1.0f, 0.6f, 1.0f) );
m_Renderer->DrawText("Waaay smaller text here!!!! (font size 30)", 600.0f, 600.0f, COORD_MODE_ABS);
m_Renderer->SetFont( m_ConsoleFont30, glm::vec4(0.0f, 0.0f, 1.0f, 0.5f) );
m_Renderer->DrawText("Waaay smaller text here!!!! (font size 30)", 605.0f, 605.0f, COORD_MODE_ABS);

m_Renderer->End2D(); // Stop 2D mode. Unbind 2d offscreen framebuffer.
```

- Every call to either `DrawText` or `DrawBox` issue a draw call. So only text glyphs get batched for one string but not
  multiple calls to `DrawText`. For `DrawBox` it is even worse because no shape gets batched even if they share the
same color. I used to have this batching in place but it makes the renderer much more complicated and I did not see
a performance improvement over having single draw calls. However, this will change if you do 1000s of shape calls! So
depending on how crazy we get with 2d rendering we might have to tackle the render backend again. I expect the backend to change anyways, so I didn't see a good reason to overcomplicate things now.

- Also the OpenGL Debug Callback was integrated to get information of API misusage.

There are some issues with alpha blending! Those become apparent when you draw shapes and/or text that is not
fully opaque!

Drawing transparent fonts or boxes  on top of other shapes/text that has already been drawn fully opaque make them translucent and therefore you can see through them into the 3d scene that should not be visible!
renderdoc_font_blending_2

### CODE REVIEW

**Review by Cameramorphic — COMMENTED** [review]
I tested the 2d rendering locally, worked great (apart from the already mentioned blending limitations). The alpha blending limitations shouldn't be an issue for now though, as we'll probably only start off with opaque text on transparent backgrounds anyway.

There are quite a few `TODO`s and `FIXME`s and commented out code, where I'm not always sure to what extent those are meant for later, or just remainders from debugging (in which case they should be removed). I only commented on some of them.
I ignored the white-space changes, that's what we'll have linting for ;)

**Comment by Cameramorphic** [comment]
It would be nice to have access to some of the horizontal font/glyph metrics for calculating placement / line lengths.
If that's a property of the font, it would only make sense for mono-space fonts though (where the width/advance should hopefully be the same for every character).
Not sure what would make sense to expose, maybe the *font bounding box* or the *advance width*/*left side bearing*?

It could also be interesting to use the vertical *line gap* metric for default line heights, though that isn't really necessary because it is easy to calculate an own line height based on the chosen font size.

File: r_gl.cpp [hunk 1]
```diff
@@ -68,7 +136,7 @@ bool GLRender::Init(void)
         SDL_WINDOWPOS_UNDEFINED,
         WINDOW_WIDTH,
         WINDOW_HEIGHT,
-        SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI
+        SDL_WINDOW_OPENGL | SDL_WINDOW_ALLOW_HIGHDPI
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_gl.cpp:139]: Good question! Forgot to mention that. On resize the FBO's textures have to be recreated since I am going to use a newer way of texture layout for the offscreen framebuffers in the upcoming iteration that are faster than regular textures but don't allow resizing by default. This is just a precaution so that resizing will be handled properly in the future.

File: r_gl.cpp [hunk 2]
```diff
@@ -189,6 +271,7 @@ int GLRender::RegisterModel(HKD_Model* model)
 
     for (int i = 0; i < model->meshes.size(); i++) {
         HKD_Mesh* mesh = &model->meshes[i];
+        //GLTexture* texture = (GLTexture*)m_TextureManager->CreateTexture("fonts/HackNerdFont-Bold.ttf");
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_gl.cpp:274]: It was! Removed it.

File: r_gl.cpp [hunk 3]
```diff
@@ -614,13 +963,67 @@ void GLRender::RenderColliders(Camera* camera, HKD_Model** models, uint32_t numM
...
+
+    // Render ImGui Elements ontop of everything.
     ImGui::Render();
     ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
 
     SDL_GL_SwapWindow(m_Window);
 
     m_ImPrimitiveBatch->Reset();
     m_ImPrimitiveBatchIndexed->Reset();
+    m_FontBatch->Reset();
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_gl.cpp:1025]: Good catch. Actually, in *theory* even the `m_FontBatch` should not have to be reset here either because it is expected to `Flush*` both shapes and fonts earlier to `RenderEnd`. However, not resetting them here could potentially crash the program if the user *never* calls either `Flush*` function (it would crash with a good errormessage however that the batch has run full and no more memory is available). 
Long story short, the `Reset` function are basically free, so no harm in calling them here as well, just to be safe!
Thanks.

File: r_gl_fbo.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
+    if (status != GL_FRAMEBUFFER_COMPLETE) {
+        printf("Framebuffer not complete: 0x%x\n", status);
+    } else {
+        printf("Framebuffer is complete\n");
+    }
+   
+    Unbind();
+
+    Bind(); // TODO: (Michael): Debugging Test. Delete!
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_gl_fbo.cpp:48]: Ugh. Important one. Thanks.

File: r_model.cpp [hunk 5]
```diff
@@ -48,6 +48,9 @@ HKD_Model CreateModelFromIQM(IQMModel* model)
         else {
             mesh.isTextured = true;
         }
+        
+        // Just load the font here so that it gets bound. TODO: Remove commented code below later.
+        // mesh.textureFileName = "fonts/HackNerdFont-Bold.ttf"; // iqmMesh->material;
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_model.cpp:53]: Removed it. Also was a way to see fonts in renderdoc quickly ;)

File: r_texture_mgr.cpp [hunk 6]
```diff
@@ -16,3 +17,12 @@ ITexture* TextureManager::CreateTexture(std::string filename)
...
 }
+
+ITexture* TextureManager::CreateTexture(CFont font)
+{
+	if (m_NameToTexture.contains(font.m_Filename)) {
+		return m_NameToTexture.at(font.m_Filename);
+	}
+
+	
+}*/
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_texture_mgr.cpp:28]: Oh yeah... Ignore that file for now. It is useless right now. I will remove it now. Bene also got confused about this, and rightfully so. It was intended to abstract the texture manager from a concrete API but that was just premature optimization on my side. Sorry.

File: r_gl_fbo.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+#include "r_gl_rendertexture.h"
+
+// Not usable in this state! Call any of the non-default Ctors to
+// get the FBO into valid state.
+CglFBO::CglFBO() { 
+    m_Width = 0;
+    m_Height = 0;
+    m_hFBO = 9999; // TODO: (Michael): Can the handle be 0? Check GL docs!
+}
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply r_gl_fbo.cpp:15]: Sometimes it can be useful to have object in uninitialized state. C++'s RAII is actually a dumb idea. But I agree that we could just delete it and set those values in the header file. Not sure...

File: game.cpp [hunk 8]
```diff
@@ -251,50 +258,102 @@ bool Game::RunFrame(double dt)
...
+        m_Renderer->RenderColliders(&m_FollowCamera, playerColliderModel, 1);
+
+        m_Renderer->End3D();
     }
 
-	// Render Player's ellipsoid collider
-    m_Renderer->SetActiveCamera(&m_FollowCamera);
-    HKD_Model* playerColliderModel[] = {&m_Player};
-    m_Renderer->RenderColliders(&m_FollowCamera, playerColliderModel, 1);
+    #if 1 // Toggle 2D Font/Box renderingtest
```
[the remark this answers is not in the document]
Comment (michaeleggers) [reply game.cpp:311]: Yup I'd like to keep it here so on merging I can glue everything together by just deleting some lines that result in a conflict.

**Comment by michaeleggers** [comment]
> It would be nice to have access to some of the horizontal font/glyph metrics for calculating placement / line lengths. If that's a property of the `CFont`, it would only make sense for mono-space fonts though (where the width/advance should hopefully be the same for every character). Not sure what would make sense to expose, maybe the _font bounding box_ or the _advance width_/_left side bearing_?
> 
> It could also be interesting to use the vertical _line gap_ metric for default line heights, though that isn't really necessary because it is easy to calculate an own line height based on the chosen font size.

Totally reasonable. At the moment the API for font (shapes) renderin is pretty crude. If it is ok for you, I'd like to create a dedicated story for that and look at this for the next, or possibly even 3rd milestone? Until we have most of the components in place, I think it is ok for things to look a bit ugly ;)

**Review by Cameramorphic — APPROVED** [review]
